Better choice of offset-text.

anntzer opened the pull request:
The axis offset text is chosen as follows:

xlims => offsettext
123, 189 => 0
12341, 12349 => 12340
99999.5, 100010.5 => 100000 # (also a test for #5780)
99990.5, 100000.5 => 100000
1233999, 1234001 => 1234000

(and the same for negative limits).

See #5755.

WeatherGod commented:
I have some test cases from a while back when I last tried to come up with a better offset text logic. These test cases really helped explore the possible edge cases in that attempt. You might find it useful (the test data, that is, not the code): https://gist.github.com/WeatherGod/272f4022bf7a8ca12ff4

anntzer commented:
Thanks @WeatherGod, I incorporated them in the test suite and slightly changed the algorithm.  Now, the main condition of whether to use the offset is whether there are two common significant digits at the beginning of every label (allowing for negative labels as discussed above, I think that's the main questionable choice remaining).

anntzer commented:
(hum, the failing test seems to work for me (and doesn't involve offset texts))

tacaswell commented on lib/matplotlib/tests/test_ticker.py:
```diff
@@ -159,6 +159,49 @@ def test_SymmetricalLogLocator_set_params():
     nose.tools.assert_equal(sym.numticks, 8)
 
 
+def test_ScalarFormatter_offset_value():
```
This needs and `@cleanup` decorator

anntzer replied:
As it is, the `@cleanup` decorator doesn't support generative tests.  I'll write an patch for that first.

tacaswell replied:
Awesome.  That is something that has driven me crazy a couple of times.  I have dealt with it by putting the decorator on the called function and creating all of the figures/axes in that function.

anntzer replied:
See #5809.

WeatherGod commented:
Huh, never seen this error before: "RuntimeError: In set_text: could not load glyph". Also, this branch needs rebasing.

anntzer commented:
Rebased.  I cannot reproduce the glyph-loading errors locally.

QuLogic commented on lib/matplotlib/ticker.py:
```diff
@@ -548,33 +548,57 @@ def set_locs(self, locs):
+        abs_min, abs_max = sorted([abs(float(lmin)), abs(float(lmax))])
+        # Only use offset if there are at least two ticks and every tick has
+        # the same sign.
+        if lmin == lmax or lmin <= 0 <= lmax:
```
Can move this section above the absolute values to short-circuit a little earlier.

QuLogic commented on lib/matplotlib/tests/test_ticker.py:
```diff
@@ -159,6 +159,50 @@ def test_SymmetricalLogLocator_set_params():
+                 (-12335.3, 12335.3, 0)]
+
+    for left, right, offset in test_data:
+        yield check_offset_for, left, right, offset
```
I _think_ all but two test cases are `left < right`; would it also make sense to yield in the reverse order? Also, I don't think there are any tests for `left == right` (though I don't see that won't work correctly.)

anntzer replied:
I don't really think I need to support left == right because I don't see how this can ever happen.  Other issues handled in new (rebased) commit.

QuLogic replied:
Well, your code does check for the `left == right` case, so it's good so codify what the expected behaviour is in that case. I think the default locators would avoid this situation, but I'm not sure about user-created locators.

anntzer replied:
Sure.  I'll wait for #6022 to be merged in so that the tests are actuall run.

anntzer replied:
updated.

QuLogic commented on lib/matplotlib/ticker.py:
```diff
@@ -548,33 +548,57 @@ def set_locs(self, locs):
+            # Handle the case of straddling a multiple of a large power of ten
+            # (relative to the span).
+            # What is the smallest power of ten such that abs_min and abs_max
+            # at most 1 apart?
```
s/at/are at/

QuLogic commented on lib/matplotlib/ticker.py:
```diff
@@ -548,33 +548,57 @@ def set_locs(self, locs):
+            # (relative to the span).
+            # What is the smallest power of ten such that abs_min and abs_max
+            # at most 1 apart?
+            oom = math.ceil(math.log10(abs_max))
```
Instead of re-calculating this (which is maybe expensive?), can you just increase `oom` until it meets the opposite condition?

QuLogic replied:
Assuming you won't get an infinite loop somehow...

VincentVandalon commented on lib/matplotlib/ticker.py:
```diff
@@ -548,33 +548,53 @@ def set_locs(self, locs):
 
-    def _set_offset(self, range):
-        # offset of 20,001 is 20,000, for example
+    def _compute_offset(self):
```
I fully agree that renaming this method was really needed; the new name actually reflects the purpose of the function.

tacaswell replied:
Contra my reluctance to change the public API at all, changing anything with a leading `_` is fair game.

efiring commented:
This is failing the two tests in which `left==right`.  It's not clear to me whether the problem is in the algorithm, or in the tests.

anntzer commented:
The test failure was probably masked at some point by the misuse of a generative test.  Anyways, this comes down to a policy choice for the left == right case.  I guess that "by continuity", we should just let the offset be equal to the single value in that case (say the values were x-eps and x+eps, then it would be normal for x (rounded to ~eps) to be the offset, so we should keep that as eps->0).
Thoughts?

efiring commented:
The present algorithm is giving `(1, 1, 1)` and `(123, 123, 120)` as left, right, offset.  Nonsingular is expanding the *actual* left, right ranges to `(0.999, 1.001)` and `(122.877, 123.123)`.  So the test cases with `left==right` are not really special cases.  Either we like what the algorithm does, or we don't.  I still don't completely understand what the algorithm's criteria are, but the results at least look reasonable, so I would be inclined to just change the tests to match.  Maybe the tests should actually be changed to use the values provided by the present implementation of nonsingular rather than feeding in equal numbers, since otherwise a change to nonsingular would cause this test to fail for reasons not related to the algorithm it is testing.

anntzer commented:
OK, now I remember what I did: an offset text is used if it "saves" at least two significant digits, i.e. if the length of the (low, high) range is no more than a hundredth of the largest power of ten below high.  So yes, the (1, 1, 1) and (123, 123, 120) outputs are expected given the effect of nonsingular.  I rewrote the tests (still using equal left and right for now because using different left and right is basically already tested) and rebased on master.

WeatherGod commented:
This is looking good to me. Were there any remaining concerns? Do we need to update any documentation about the improved offset text algorithm?

anntzer commented:
I can add a snippet like "The default offset-text choice was changed to only use significant digits that are common to all ticks (e.g. 1231..1239 -> 1230, instead of 1231), except when they straddle a relatively large multiple of a power of ten in which case that multiple is chosen (e.g. 1999..2001->2000)."  Looks good?

efiring commented:
@anntzer Yes, adding that explanation would be good.

anntzer commented:
Done.

dopplershift commented:
Looks like the only failure on AppVeyor was spurious. Anyone with proper permissions wanna restart the appveyor run so this can go green and be merged?

efiring commented:
I don't see a way to restart just one appveyor test, so I triggered the whole set of checks.

dopplershift commented:
If you have permissions, you can log into app-veyor and re-run the build. Unfortunately, it doesn't key off of GitHub permissions; those permissions are managed manually by whoever owns the appveyor project (looks like @mdboom )

efiring commented:
Backported to v2.x as beb08c8.
